src/controller.rs

```diff
@@ -46,46 +46,136 @@ struct Context {
+    error: &Error,
+    _ctx: Arc<Context>,
+) -> Action {
     let name = app_instance.name_any();
     warn!(?name, %error, "reconcile failed");
     // TODO(mkm): make error requeue duration configurable
     Action::requeue(Duration::from_secs(5))
 }
 
+fn error_policy_config_map(
+    config_map: Arc<ConfigMap>,
+    error: &Error,
+    _ctx: Arc<Context>,
+) -> Action {
+    let config = &config_map.as_ref().data.as_ref().unwrap()["app-instance"];
+    let ai: Result<AppInstance, serde_yaml::Error> = serde_yaml::from_str(config);
+    match ai {
+        Ok(ai) => error_policy_app_instance(Arc::new(ai), error, _ctx),
+        Err(_) => {
+            error!(%error, "failed to convert configmap to appinstance");
+            std::process::exit(1);
+        }
+    }
+}
+
+async fn reconcile_app_instance(
+    app_instance: Arc<AppInstance>,
+    ctx: Arc<Context>,
+) -> Result<Action> {
+    AppInstanceLike::from(app_instance).reconcile(ctx).await
+}
+
+async fn reconcile_config_map(config_map: Arc<ConfigMap>, ctx: Arc<Context>) -> Result<Action> {
+    let cm_name = config_map.name_any();
+    if ctx
+        .as_ref()
+        .config_map_name
+        .as_ref()
+        .is_some_and(|x| x.as_str() == cm_name)
+    {
+        let ai = AppInstanceLike::from_config_map(config_map, "app-instance");
+        match ai {
+            Ok(ai) => ai.reconcile(ctx).await,
+            Err(error) => {
+                error!(%error, "failed to convert configmap to appinstance");
+                std::process::exit(1);
+            }
+        }
+    } else {
+        Ok(Action::requeue(Duration::from_secs(5)))
+    }
+}
+
 pub async fn run(
     client: Client,
     kubecfg_image: String,
     kubit_image: String,
     only_paused: bool,
+    config_map_name: Option<String>,
+    watched_namespace: Option<String>,
 ) -> Result<()> {
-    let docs = Api::<AppInstance>::all(client.clone());
-    if let Err(e) = docs.list(&ListParams::default().limit(1)).await {
-        error!("CRD is not queryable; {e:?}. Is the CRD installed?");
-        std::process::exit(1);
-    }
+    let namespace = watched_namespace.as_deref();
+
+    let jobs = if let Some(ns) = namespace {
+        Api::<Job>::namespaced(client.clone(), ns)
+    } else {
+        Api::<Job>::all(client.clone())
+    };
 
     info!("running kubit manager");
-    let jobs = Api::<Job>::all(client.clone());
-    Controller::new(docs, watcher::Config::default().any_semantic())
-        .shutdown_on_signal()
-        .owns(jobs, watcher::Config::default().any_semantic())
-        .run(
-            reconcile,
-            error_policy,
-            Arc::new(Context {
-                client,
-                kubecfg_image,
-                kubit_image,
-                only_paused,
-            }),
-        )
-        .filter_map(|x| async move { std::result::Result::ok(x) })
-        .for_each(|_| futures::future::ready(()))
-        .await;
+    if config_map_name.is_none() {
```

Comment:
I wonder whether a small couple of extra `info!` calls in here is useful for a mention to which "mode" the controller is running in?

I.e. if you see it running in configmap mode but you have the CRD installed, it's easier to spot right at the time of startup (or in the case of a customer dumping logs and we can tie things together etc)